pkg/client/query/cache/config.go

```diff
@@ -0,0 +1,69 @@
+package cache
+
+import (
+	"time"
+)
+
+// EvictionPolicy determines how items are removed when number of keys in the cache reaches MaxKeys.
+type EvictionPolicy int64
+
+const (
+	FirstInFirstOut = EvictionPolicy(iota)
+	LeastRecentlyUsed
+	LeastFrequentlyUsed
+)
+
+// queryCacheConfig is the configuration for query caches. It is intended to be
+// configured via QueryCacheOptionFn functions.
+type queryCacheConfig struct {
+	// MaxKeys is the maximum number of items (key/value pairs) the cache can
+	// hold before it starts evicting.
+	MaxKeys        int64
+	EvictionPolicy EvictionPolicy
+	// TTL is how long items should remain in the cache. Items older than the TTL
+	// MAY not be evicted but SHOULD not be considered as cache hits.
+	TTL time.Duration
+
+	// historical determines whether the cache will cache a single value for each
+	// key (false), or whether it will cache a history of values for each key (true).
+	historical bool
+	// pruneOlderThan is the number of past blocks for which to keep historical
+	// values. If 0, no historical pruning is performed. It only applies when
+	// historical is true.
+	pruneOlderThan int64
+}
+
+// QueryCacheOptionFn is a function which receives a queryCacheConfig for configuration.
+type QueryCacheOptionFn func(*queryCacheConfig)
+
+// WithHistoricalMode enables historical caching with the given pruneOlderThan
+// configuration; if 0, no historical pruning is performed.
+func WithHistoricalMode(pruneOlderThan int64) QueryCacheOptionFn {
+	return func(cfg *queryCacheConfig) {
+		cfg.historical = true
+		cfg.pruneOlderThan = pruneOlderThan
+	}
+}
+
+// WithMaxKeys sets the maximum number of distinct key/value pairs the cache will
```

Comment:
TOL: MaxEntries sounds pretty bad but that's what this is, no? Just curious what you think.

Reply:
I considered `MaxEntries` but didn't feel like it was as clear as `MaxKeys`. It would be more idiomatic if we were in a different language (e.g. javascript, python, ruby, C#, etc.). :sweat_smile: